src/gt4py/next/iterator/transforms/collapse_tuple.py

```diff
@@ -312,6 +358,73 @@ def transform_propagate_to_if_on_tuples(self, node: ir.FunCall, **kwargs) -> Opt
                     return im.if_(cond, new_true_branch, new_false_branch)
         return None
 
+    def transform_propagate_to_if_on_tuples_cps(
+        self, node: ir.FunCall, **kwargs
+    ) -> Optional[ir.Node]:
+        if not cpm.is_call_to(node, "if_"):
+            for i, arg in enumerate(node.args):
+                if cpm.is_call_to(arg, "if_"):
+                    itir_type_inference.reinfer(arg)
+                    if not any(isinstance(branch.type, ts.TupleType) for branch in arg.args[1:]):
+                        continue
+
+                    cond, true_branch, false_branch = arg.args
+                    tuple_type: ts.TupleType = true_branch.type  # type: ignore[assignment]  # type ensured above
+                    tuple_len = len(tuple_type.types)
+                    itir_type_inference.reinfer(node)
+                    assert node.type
+
+                    # transform function into continuation-passing-style
+                    f_type = ts.FunctionType(
+                        pos_only_args=tuple_type.types,
+                        pos_or_kw_args={},
+                        kw_only_args={},
+                        returns=node.type,
+                    )
+                    f_params = [
+                        im.sym(self.uids.sequential_id(prefix="__ct_el_cps"), type_)
+                        for type_ in tuple_type.types
+                    ]
+                    f_args = [im.ref(param.id, param.type) for param in f_params]
+                    f_body = _with_altered_arg(node, i, im.make_tuple(*f_args))
+                    # simplify, e.g., inline trivial make_tuple args
+                    new_f_body = self.fp_transform(f_body, **kwargs)
```

Comment:
In this recursion, we handle all the other args after the current one? I am confused...